### PR TITLE
Fix to #10723 - Query: add correlated collection optimization support for async queries

### PR DESCRIPTION
**Pull request by maumar**
Adding support for async, porting GearsOfWars correlated collection tests to async for coverage.

### CODE REVIEW

File: src/EFCore/Query/ExpressionVisitors/Internal/CorrelatedCollectionOptimizingVisitor.cs [hunk 1]
```diff
@@ -225,9 +232,18 @@ var joinQuerySourceReferenceExpression
 
             if (navigation.ForeignKey.DeclaringEntityType.ClrType == collectionQueryModel.SelectClause.Selector.Type)
             {
-                correlateSubqueryMethod = _correlateSubqueryMethodInfo.MakeGenericMethod(
-                    collectionQueryModel.SelectClause.Selector.Type,
-                    navigation.GetCollectionAccessor().CollectionType);
+                if (_queryCompilationContext.AsyncQuery)
```
Comment (anpete) [inline src/EFCore/Query/ExpressionVisitors/Internal/CorrelatedCollectionOptimizingVisitor.cs]: Use this test to resolve the open MethodInfo before calling MakeGenericMethod.

File: src/EFCore/Query/QueryCompilationContext.cs [hunk 2]
```diff
@@ -61,6 +61,7 @@ public class QueryCompilationContext
             LinqOperatorProvider = linqOperatorProvider;
             ContextType = dependencies.CurrentContext.Context.GetType();
             TrackQueryResults = trackQueryResults;
+            AsyncQuery = LinqOperatorProvider.Select.ReturnType.GetGenericTypeDefinition() == typeof(IAsyncEnumerable<>);
```
Comment (maumar) [inline src/EFCore/Query/QueryCompilationContext.cs]: is there better way for determining sync vs async query?

File: src/EFCore/Query/ExpressionVisitors/Internal/CorrelatedCollectionOptimizingVisitor.cs [hunk 3]
```diff
@@ -241,9 +257,18 @@ var joinQuerySourceReferenceExpression
                 var resultCollectionType = typeof(List<>).MakeGenericType(collectionQueryModel.SelectClause.Selector.Type);
                 var resultCollectionCtor = resultCollectionType.GetTypeInfo().GetDeclaredConstructor(new Type[] { });
 
-                correlateSubqueryMethod = _correlateSubqueryMethodInfo.MakeGenericMethod(
-                    collectionQueryModel.SelectClause.Selector.Type,
-                    typeof(List<>).MakeGenericType(collectionQueryModel.SelectClause.Selector.Type));
+                if (_queryCompilationContext.AsyncQuery)
```
Comment (anpete) [inline src/EFCore/Query/ExpressionVisitors/Internal/CorrelatedCollectionOptimizingVisitor.cs]: As above.

File: src/EFCore/Query/ExpressionVisitors/Internal/CorrelatedCollectionOptimizingVisitor.cs [hunk 4]
```diff
@@ -273,6 +298,15 @@ var joinQuerySourceReferenceExpression
                 ? typeof(IOrderedEnumerable<>).MakeGenericType(collectionQueryModel.SelectClause.Selector.Type)
                 : typeof(IEnumerable<>).MakeGenericType(collectionQueryModel.SelectClause.Selector.Type);
 
+            var lambda = (Expression)Expression.Lambda(new SubQueryExpression(collectionQueryModel));
+            if (_queryCompilationContext.AsyncQuery)
+            {
+                lambda = Expression.Convert(
+                    lambda,
+                    typeof(Func<>).MakeGenericType(
+                        typeof(IAsyncEnumerable<>).MakeGenericType(collectionQueryModel.SelectClause.Selector.Type)));
```
Comment (anpete) [inline src/EFCore/Query/ExpressionVisitors/Internal/CorrelatedCollectionOptimizingVisitor.cs]: Make a local for collectionQueryModel.SelectClause.Selector.Type ?

File: src/EFCore/Query/QueryCompilationContext.cs [hunk 5]
```diff
@@ -116,6 +117,11 @@ public class QueryCompilationContext
         /// </value>
         public virtual QuerySourceMapping QuerySourceMapping { get; } = new QuerySourceMapping();
 
+        /// <summary>
+        ///     Get a value indicating whether query that is being processed is asynchronous.
+        /// </summary>
+        public virtual bool AsyncQuery { get; }
```
Comment (anpete) [inline src/EFCore/Query/QueryCompilationContext.cs]: IsAsyncQuery

File: src/EFCore/Query/QueryCompilationContext.cs (hunk 2 above)
Comment (anpete) [inline src/EFCore/Query/QueryCompilationContext.cs]: Why do we need this. I think we know this already.

Comment (maumar) [reply src/EFCore/Query/QueryCompilationContext.cs]: we know on the top level but then we lose that information if child EQMV gets created from another visitor, e.g. RelationalProjectionExpressionVisitor. I guess we could flow that information to all visitors that can later create a child EQMV, but that seems more disruptive, and not sure if we can predict all places when EQMV will be created in the future

Comment (anpete) [reply src/EFCore/Query/QueryCompilationContext.cs]: OK, we usually flow it from the argument passed to EQMV.OptimizeQueryModel. We can add this to QueryCompilationContext, but just set it from EQMV.CreateAsyncQueryExecutor

File: src/EFCore/Query/ExpressionVisitors/Internal/CorrelatedCollectionOptimizingVisitor.cs [hunk 6]
```diff
@@ -37,6 +38,9 @@ public class CorrelatedCollectionOptimizingVisitor : ExpressionVisitorBase
         private static MethodInfo _correlateSubqueryMethodInfo
             = typeof(IQueryBuffer).GetMethod(nameof(IQueryBuffer.CorrelateSubquery));
 
+        private static MethodInfo _correlateSubqueryAsyncMethodInfo
```
Comment (anpete) [inline src/EFCore/Query/ExpressionVisitors/Internal/CorrelatedCollectionOptimizingVisitor.cs]: These can be readonly

File: src/EFCore/Query/ExpressionVisitors/Internal/CorrelatedCollectionOptimizingVisitor.cs [hunk 7]
```diff
@@ -285,24 +319,27 @@ var joinQuerySourceReferenceExpression
...
+            if (_queryCompilationContext.AsyncQuery)
             {
-                return
-                    Expression.Call(
-                        _queryCompilationContext.LinqOperatorProvider.ToOrdered
-                            .MakeGenericMethod(result.Type.GetSequenceType()),
-                        result);
+                var taskResultExpression = new TaskBlockingExpressionVisitor().Visit(result);
+
+                return taskResultExpression;
```
Comment (anpete) [inline src/EFCore/Query/ExpressionVisitors/Internal/CorrelatedCollectionOptimizingVisitor.cs]: I think there is dead code on line 556.

Comment (anpete) [reply src/EFCore/Query/ExpressionVisitors/Internal/CorrelatedCollectionOptimizingVisitor.cs]: LiftOrderBy had two dead args.

**Comment by anpete** [comment]
We are still buffering twice to ToList case, right?

**Comment by maumar** [comment]
yes, we still have ToList calls - tried to naively remove them, but hit some problems with type mismatch. Will remove it in separate PR

**Comment by maumar** [comment]
merged this PR instead https://github.com/aspnet/EntityFrameworkCore/pull/10755

**Pull request closed without merging**